apisix/plugins/mocking.lua

```diff
@@ -0,0 +1,230 @@
+        -- specify response json schema, if response_example is not nil, this conf will be ignore.
+        -- generate random response by json schema.
+        response_schema = { type = "object" },
+        with_mock_header = { type = "boolean", default = true }
+    },
+    anyOf = {
+        { required = { "response_example" } },
+        { required = { "response_schema" } }
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 9900,
+    name = "mocking",
+    schema = schema,
+}
+
+local function parse_content_type(content_type)
+    if not content_type then
+        return ""
+    end
+    local m = ngx_re.match(content_type, "([ -~]*);([ -~]*)")
+    if m and #m == 2 then
+        return m[1], m[2]
+    end
+    return content_type
+end
+
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+    if not ok then
+        return false, err
+    end
+
+    local typ = parse_content_type(conf.content_type)
+    local support_content_type = {
+        ["application/xml"] = true,
+        ["application/json"] = true,
+        ["text/plain"] = true,
+        ["text/html"] = true,
+        ["text/xml"] = true
+    }
+    if not support_content_type[typ] then
+        return false, "unsupported content type!"
+    end
+    return true
+end
+
+
+local function gen_string(example)
+    if example and type(example) == "string" then
+        return example
+    end
+    local n = math.random(1, 10)
+    local list = {}
+    for i = 1, n do
+        table.insert(list, string.char(math.random(97, 122)))
+    end
+    return table.concat(list)
+end
+
+
+local function gen_number(example)
+    if example and type(example) == "number" then
+        return example
+    end
+    return math.random() * 10000
+end
+
+
+local function gen_integer(example)
+    if example and type(example) == "number" then
+        return math.floor(example)
+    end
+    return math.random(1, 10000)
+end
+
+
+local function gen_boolean(example)
+    if example and type(example) == "boolean" then
+        return example
+    end
+    local r = math.random(0, 1)
+    if r == 0 then
+        return false
+    end
+    return true
+end
+
+
+local function gen_base(property)
+    local typ = string.lower(property.type)
+    local example = property.example
+    if typ == "string" then
+        return gen_string(example)
+    end
+
+    if typ == "number" then
+        return gen_number(example)
+    end
+
+    if typ == "integer" then
+        return gen_integer(example)
+    end
+
+    if typ == "boolean" then
+        return gen_boolean(example)
+    end
+    return nil
+end
+
+
+local gen_array, gen_object
+
+function gen_array(property)
+    local output = {}
+    if property.items == nil then
+        return nil
+    end
+    local v = property.items
+    local n = math.random(1, 3)
+    local typ = string.lower(v.type)
+    for i = 1, n do
+        if typ == "array" then
+            table.insert(output, gen_array(v))
```

Comment:
We can refactor this part and use a function that turns typ into the generated data.